platform/core/src/services/MeasurementService/MeasurementService.js

```diff
@@ -0,0 +1,275 @@
+      });
+    }
+  }
+
+  /**
+   * Subscribe to measurement updates.
+   *
+   * @param {string} eventName
+   * @param {Function} callback
+   * @param {string} context
+   * @return {Object} observable actions
+   */
+  subscribe(eventName, callback, context = 'all') {
+    if (this._isValidEvent(eventName)) {
+      console.warn(`Subscribing to '${eventName}' event using '${context}' context.`);
+      const listenerId = guid();
+
+      /* Create new listeners context if needed */
+      if (!this.listeners[context]) {
+        this.listeners[context] = {};
+      }
+
+      if (Array.isArray(this.listeners[context][eventName])) {
+        this.listeners[context][eventName].push({ id: listenerId, callback });
+      } else {
+        this.listeners[context][eventName] = [{ id: listenerId, callback }];
+      }
+
+      return {
+        unsubscribe: () => this._unsubscribe(eventName, listenerId, context)
+      };
+    } else {
+      throw new Error(`Event ${eventName} not supported in '${context}' context.`);
+    }
+  }
+
+  /**
+   * Unsubscribe to measurement updates.
+   *
+   * @param {string} eventName
+   * @param {string} listenerId
+   * @param {string} context
+   * @return void
+   */
+  _unsubscribe(eventName, listenerId, context) {
+    if (!this.listeners[context]) {
+      return;
+    }
+
+    const listenersOfContext = this.listeners[context][eventName];
+    if (Array.isArray(listenersOfContext)) {
+      this.listeners[context][eventName] = listenersOfContext.filter(
+        ({ id }) => id !== listenerId
+      );
+    } else {
+      this.listeners[context][eventName] = undefined;
+    }
+  }
+
+  /**
+   * Check if a given measurement data is valid.
+   *
+   * @param {MeasurementSchema} measurementData
+   * @return {boolean} measurement validation
+   */
+  _isValidMeasurement(measurementData) {
+    const MEASUREMENT_SCHEMA_KEYS = [
+      'id',
+      'sopInstanceUID',
+      'frameOfReferenceUID',
+      'referenceSeriesUID',
+      'label',
+      'description',
+      'type',
+      'unit',
+      'area', // TODO: Add concept names instead (descriptor)
+      'points',
+      'source',
+      'sourceToolType',
+    ];
+
+    Object.keys(measurementData).forEach(key => {
+      if (!MEASUREMENT_SCHEMA_KEYS.includes(key)) {
+        log.warn(`Invalid measurement key: ${key}`);
+        return false;
+      }
+    });
+
+    return true;
+  }
+
+  /**
+   * Check if a given measurement service event is valid.
+   *
+   * @param {string} eventName
+   * @return {boolean} event name validation
+   */
+  _isValidEvent(eventName) {
+    return Object.values(this.events).includes(eventName);
+  }
+
+  /**
+   * Converts object of objects to array.
+   *
+   * @return {Array} Array of objects
+   */
+  _arrayOfObjects = obj => {
+    return Object.entries(obj).map(e => ({ [e[0]]: e[1] }));
+  };
+}
+
+export default MeasurementService;
+export { EVENTS };
```

Comment:
Top level API:

- subscribe(eventName, callback, context = 'all')
- addOrUpdate(measurement, context = 'all')
- registerEvent(eventName)
- getEvents()
- getMeasurement(id, context)
- getMeasurements(context = 'all')
- VALUE_TYPES

Separate:
- EVENTS

Reply:
As events are emitted internally by the `MeasurementService`, I'm not sure when a newly registered event would be emitted or subscribed to as part of the `MeasurementService`'s normal operation.

The only dynamic event emitting I can think of, and we don't have this yet, would be scoping listening to one or more callers. IE. You only want to know if a measurement is modified by the VTK extension, or by any extension except for VTK. (No need to chase this down, just thinking out loud).

I'm not against a `registerEvent` and `getEvents`, but I'd want to know of some planned usage before we commit to them. If we don't have anything in mind, I'd rather we get rid of both and do a named export of EVENTS only.

Reply:
It look like the song and dance for using this service is:

**Push:**

- Listen for library X's created/updated/deleted events
- Map measurement to `MeasurementService`'s format
- Call `MeasurementService.addOrUpdate` with a "Measurement" and a "context"
- If creating, use the returned ID so we can update/sync in the future

**Pull:**

- Subscribe to `MeasurementService` events (should we use addEventListener?)
- OnAdd (currently only log)
- OnUpdated
    - Use `measurement.source` as an alias for tool name
    - Use mapper's `toAnnotation` on known supported type?